File: qbp-chain/sdk/go/qbp.go
// Package qbp provides the official Go SDK for interacting with a
// Quantum Blockchain Pro (QBP) node.
//
// It allows Go applications to:
//   - Connect to a QBP node via RPC
//   - Manage post-quantum accounts (ML-DSA/ML-KEM)
//   - Send transactions with ML-DSA signatures
//   - Deploy and interact with smart contracts
//   - Submit and retrieve quantum computation jobs
package qbp

import (
	"context"
	"crypto/sha3"
	"fmt"
	"math/big"

	"github.com/NikaHsaini/Quantum-Blockchain-Pro/qbp-chain/crypto/pqcrypto"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/ethclient"
	"github.com/ethereum/go-ethereum/rpc"
)

// ============================================================
// QBP Client
// ============================================================

// Client is a QBP client for interacting with a QBP node.
// It embeds an go-ethereum ethclient and adds QBP-specific functionality.
type Client struct {
	*ethclient.Client
	rpc *rpc.Client
}

// Dial connects a client to the given URL.
func Dial(rawurl string) (*Client, error) {
	rpcClient, err := rpc.Dial(rawurl)
	if err != nil {
		return nil, err
	}
	ethClient := ethclient.NewClient(rpcClient)

	return &Client{
		Client: ethClient,
		rpc:    rpcClient,
	}, nil
}

// ============================================================
// Post-Quantum Account Management
// ============================================================

// Account represents a QBP account with post-quantum keys.
type Account struct {
	Address    common.Address
	PrivateKey *pqcrypto.MLDSAPrivateKey
	PublicKey  *pqcrypto.MLDSAPublicKey
}

// NewAccount creates a new QBP account with a new ML-DSA key pair.
func NewAccount() (*Account, error) {
	keyPair, err := pqcrypto.GenerateMLDSAKeyPair()
	if err != nil {
		return nil, fmt.Errorf("qbp-sdk: failed to generate ML-DSA key pair: %w", err)
	}

	addr := keyPair.PublicKey.Address()

	return &Account{
		Address:    common.BytesToAddress(addr[:]),
		PrivateKey: keyPair.PrivateKey,
		PublicKey:  keyPair.PublicKey,
	}, nil
}

// AccountFromPrivateKey creates an Account from an existing ML-DSA private key.
func AccountFromPrivateKey(privKey *pqcrypto.MLDSAPrivateKey) *Account {
	pubKey := privKey.GetPublicKey()
	addr := pubKey.Address()

	return &Account{
		Address:    common.BytesToAddress(addr[:]),
		PrivateKey: privKey,
		PublicKey:  pubKey,
	},
}

// ============================================================
// Transaction Signing
// ============================================================

// SignTx signs a transaction with the account's ML-DSA private key.
// This is the post-quantum equivalent of types.SignTx in go-ethereum.
func SignTx(tx *Transaction, signer Signer, prv *pqcrypto.MLDSAPrivateKey) (*Transaction, error) {
	h := signer.Hash(tx)
	sig, err := pqcrypto.Sign(prv, h[:])
	if err != nil {
		return nil, err
	}

	return tx.WithSignature(signer, sig.Bytes())
}

// Transaction represents a QBP transaction.
// It is compatible with go-ethereum's types.Transaction but uses a different signature scheme.
type Transaction struct {
	// Fields compatible with go-ethereum/core/types.Transaction
	Nonce    uint64
	GasPrice *big.Int
	GasLimit uint64
	To       *common.Address
	Value    *big.Int
	Data     []byte

	// QBP-specific post-quantum signature
	// V, R, S are replaced by a single ML-DSA signature
	PQSignature []byte
}

// WithSignature returns a new transaction with the given signature.
func (tx *Transaction) WithSignature(signer Signer, sig []byte) (*Transaction, error) {
	cpy := *tx
	cpy.PQSignature = sig
	return &cpy, nil
}

// Hash computes the hash of the transaction to be signed.
func (tx *Transaction) Hash() common.Hash {
	h := sha3.New256()
	// Simplified hashing for SDK example
	h.Write(tx.Data)
	return common.BytesToHash(h.Sum(nil))
}

// Signer is an interface for signing transactions.
type Signer interface {
	Hash(*Transaction) common.Hash
}

// ============================================================
// Quantum Oracle Interaction
// ============================================================

// QuantumJob represents a quantum computation job for submission.
type QuantumJob struct {
	NumQubits   uint64
	CircuitHash [32]byte
	Deadline    *big.Int
	Reward      *big.Int
}

// SubmitQuantumJob submits a quantum computation job to the QuantumOracle contract.
func (c *Client) SubmitQuantumJob(ctx context.Context, job *QuantumJob, account *Account) (*Transaction, error) {
	// In a full implementation, this would:
	// 1. ABI-encode the call to QuantumOracle.submitJob()
	// 2. Create a new transaction with the encoded data
	// 3. Sign the transaction with the account's ML-DSA key
	// 4. Send the raw transaction

	fmt.Printf("Submitting quantum job with %d qubits...\n", job.NumQubits)

	// Create a dummy transaction for demonstration
	tx := &Transaction{
		Nonce:    0, // Should be retrieved from the network
		To:       &common.Address{}, // QuantumOracle contract address
		Value:    job.Reward,
		GasLimit: 300000,
		GasPrice: big.NewInt(20000000000), // 20 Gwei
		Data:     []byte("submitJob_encoded_data"),
	}

	// Sign the transaction
	signer := &QBPSigner{}
	signedTx, err := SignTx(tx, signer, account.PrivateKey)
	if err != nil {
		return nil, err
	}

	fmt.Printf("Transaction signed with ML-DSA signature of length %d\n", len(signedTx.PQSignature))

	// Send transaction (simulation)
	// err = c.SendTransaction(ctx, signedTx)
	// if err != nil { ... }

	return signedTx, nil
}

// GetQuantumJobResult retrieves the result of a quantum job.
func (c *Client) GetQuantumJobResult(ctx context.Context, jobID [32]byte) (string, error) {
	var result string
	// In a full implementation, this would call the QuantumOracle.jobs() view function.
	err := c.rpc.CallContext(ctx, &result, "qbp_getQuantumJobResult", jobID)
	if err != nil {
		return "", err
	}
	return result, nil
}

// QBPSigner implements the Signer interface for QBP transactions.
type QBPSigner struct{}

// Hash returns the hash to be signed for a QBP transaction.
func (s *QBPSigner) Hash(tx *Transaction) common.Hash {
	return tx.Hash()
}

// ============================================================
// Example Usage
// ============================================================

// Example demonstrates how to use the QBP Go SDK.
func Example() {
	// 1. Connect to a QBP node
	client, err := Dial("http://localhost:8545")
	if err != nil {
		fmt.Printf("Failed to connect to QBP node: %v\n", err)
		return
	}

	// 2. Create a new post-quantum account
	account, err := NewAccount()
	if err != nil {
		fmt.Printf("Failed to create account: %v\n", err)
		return
	}
	fmt.Printf("New QBP Account: %s\n", account.Address.Hex())

	// 3. Get the latest block number
	block, err := client.BlockByNumber(context.Background(), nil)
	if err != nil {
		fmt.Printf("Failed to get latest block: %v\n", err)
		return
	}
	fmt.Printf("Latest Block: %d\n", block.Number().Uint64())

	// 4. Submit a quantum job
	job := &QuantumJob{
		NumQubits:   8,
		CircuitHash: [32]byte{1, 2, 3},
		Deadline:    big.NewInt(int64(block.NumberU64() + 1000)),
		Reward:      big.NewInt(100 * 1e18), // 100 QBP
	}

	tx, err := client.SubmitQuantumJob(context.Background(), job, account)
	if err != nil {
		fmt.Printf("Failed to submit quantum job: %v\n", err)
		return
	}

	fmt.Printf("Submitted quantum job with transaction hash: %s\n", tx.Hash().Hex())
}

File: qbp-chain/sdk/js/qbp.js
/**
 * @file qbp.js
 * @author Quantum Blockchain Pro Team
 * @description Official JavaScript SDK for interacting with a Quantum Blockchain Pro (QBP) node.
 *
 * This SDK provides a convenient interface for web applications and Node.js services to:
 *   - Connect to a QBP node via WebSockets or HTTP
 *   - Manage post-quantum accounts (ML-DSA/ML-KEM)
 *   - Sign and send transactions with ML-DSA signatures
 *   - Interact with the Quantum Oracle and QMaaS marketplace
 *   - Listen for on-chain quantum events
 *
 * It is built on top of ethers.js, extending it with QBP-specific features.
 */

const { ethers } = require("ethers");
const pqcrypto = require("./pqcrypto_lib"); // Placeholder for post-quantum crypto library

// ============================================================
// QBP Provider
// ============================================================

/**
 * QBPProvider extends ethers.JsonRpcProvider to add QBP-specific RPC methods.
 */
class QBPProvider extends ethers.JsonRpcProvider {
    constructor(url) {
        super(url);
    }

    /**
     * Get the result of a quantum computation job.
     * @param {string} jobId - The ID of the job (bytes32 hex string).
     * @returns {Promise<object>} The quantum job result.
     */
    async getQuantumJobResult(jobId) {
        return this.send("qbp_getQuantumJobResult", [jobId]);
    }

    /**
     * Get the list of active quantum miners.
     * @returns {Promise<string[]>} An array of miner addresses.
     */
    async getActiveMiners() {
        return this.send("qbp_getActiveMiners", []);
    }
}

// ============================================================
// QBP Wallet (Post-Quantum)
// ============================================================

/**
 * QBPWallet represents a QBP account with a post-quantum (ML-DSA) key pair.
 * It extends ethers.Wallet to override signing methods.
 */
class QBPWallet extends ethers.Wallet {
    /**
     * Creates a new QBPWallet instance.
     * @param {object} privateKey - The ML-DSA private key object.
     * @param {QBPProvider} provider - The QBP provider.
     */
    constructor(privateKey, provider) {
        // Pass a dummy private key to the parent constructor
        super(ethers.hexlify(ethers.randomBytes(32)), provider);

        this.mldsaPrivateKey = privateKey;
        this.mldsaPublicKey = pqcrypto.getPublicKey(privateKey);
        this.address = this.getAddressFromPublicKey(this.mldsaPublicKey);
    }

    /**
     * Creates a new random QBP wallet.
     * @param {QBPProvider} provider - The QBP provider.
     * @returns {QBPWallet} A new wallet instance.
     */
    static createRandom(provider) {
        const privateKey = pqcrypto.generateMLDSAKeyPair();
        return new QBPWallet(privateKey, provider);
    }

    /**
     * Derives the QBP address from an ML-DSA public key.
     * @param {object} publicKey - The ML-DSA public key.
     * @returns {string} The QBP address.
     */
    getAddressFromPublicKey(publicKey) {
        const pkBytes = pqcrypto.serializePublicKey(publicKey);
        const hash = ethers.keccak256(pkBytes);
        return ethers.getAddress(ethers.dataSlice(hash, 12)); // Last 20 bytes
    }

    /**
     * Signs a transaction with the wallet's ML-DSA private key.
     * @param {object} tx - The transaction object.
     * @returns {Promise<string>} The signed transaction hex string.
     */
    async signTransaction(tx) {
        // In QBP, the signature replaces the v, r, s fields.
        // The transaction hash to be signed is computed differently.
        const txHash = this.computeTxHash(tx);

        const signature = pqcrypto.sign(this.mldsaPrivateKey, txHash);
        const serializedSignature = pqcrypto.serializeSignature(signature);

        // Add the post-quantum signature to the transaction
        tx.pqSignature = ethers.hexlify(serializedSignature);

        // Serialize the transaction with the PQ signature
        return this.serializeTx(tx);
    }

    /**
     * Computes the hash of a transaction to be signed.
     * @param {object} tx - The transaction object.
     * @returns {Uint8Array} The 32-byte transaction hash.
     */
    computeTxHash(tx) {
        // Simplified hashing for example
        const to = tx.to ? ethers.getAddress(tx.to) : "0x";
        const data = tx.data ? ethers.hexlify(tx.data) : "0x";
        const value = tx.value ? ethers.toBigInt(tx.value) : 0n;
        const nonce = tx.nonce ? tx.nonce : 0;

        const encoded = ethers.solidityPacked(
            ["uint256", "address", "uint256", "bytes"],
            [nonce, to, value, data]
        );

        return ethers.keccak256(encoded);
    }

    /**
     * Serializes a QBP transaction including the post-quantum signature.
     * @param {object} tx - The transaction object.
     * @returns {string} The serialized transaction hex string.
     */
    serializeTx(tx) {
        // In a real implementation, this would use RLP encoding
        // with a custom transaction type for QBP.
        return JSON.stringify(tx);
    }
}

// ============================================================
// Quantum Oracle Interaction
// ============================================================

const quantumOracleAbi = [
    "function submitJob(uint256 numQubits, bytes32 circuitHash, uint256 deadline) external payable returns (bytes32 jobId)",
    "function groverSearch(uint256 numQubits, uint256 targetState) external returns (uint256 foundState)",
    "function verifyPostQuantumSignature(bytes32 messageHash, bytes calldata publicKey, bytes calldata signature) external returns (bool isValid)",
    "event QuantumJobSubmitted(bytes32 indexed jobId, address indexed submitter, uint256 reward)",
];

/**
 * QuantumOracle provides an interface to the QuantumOracle smart contract.
 */
class QuantumOracle {
    /**
     * Creates a new QuantumOracle instance.
     * @param {string} address - The contract address.
     * @param {QBPWallet} wallet - The QBP wallet to use for transactions.
     */
    constructor(address, wallet) {
        this.contract = new ethers.Contract(address, quantumOracleAbi, wallet);
    }

    /**
     * Submits a quantum computation job.
     * @param {number} numQubits - Number of qubits.
     * @param {string} circuitHash - Hash of the circuit definition (bytes32 hex).
     * @param {number} deadline - Block number deadline.
     * @param {string} reward - Reward in QBP (e.g., "100.0").
     * @returns {Promise<ethers.TransactionResponse>} The transaction response.
     */
    async submitJob(numQubits, circuitHash, deadline, reward) {
        const rewardWei = ethers.parseEther(reward);
        return this.contract.submitJob(numQubits, circuitHash, deadline, {
            value: rewardWei,
        });
    }

    /**
     * Executes Grover's search algorithm on-chain.
     * @param {number} numQubits - Number of qubits.
     * @param {number} targetState - The state to search for.
     * @returns {Promise<bigint>} The found state.
     */
    async groverSearch(numQubits, targetState) {
        return this.contract.groverSearch(numQubits, targetState);
    }

    /**
     * Verifies an ML-DSA signature on-chain.
     * @param {string} messageHash - The hash of the message (bytes32 hex).
     * @param {Uint8Array} publicKey - The ML-DSA public key bytes.
     * @param {Uint8Array} signature - The ML-DSA signature bytes.
     * @returns {Promise<boolean>} True if the signature is valid.
     */
    async verifyPostQuantumSignature(messageHash, publicKey, signature) {
        return this.contract.verifyPostQuantumSignature(
            messageHash,
            ethers.hexlify(publicKey),
            ethers.hexlify(signature)
        );
    }
}

// ============================================================
// Example Usage
// ============================================================

async function main() {
    console.log("Connecting to QBP node...");
    const provider = new QBPProvider("http://localhost:8545");

    console.log("Creating a new post-quantum wallet...");
    const wallet = QBPWallet.createRandom(provider);
    console.log("Wallet Address:", wallet.address);

    const balance = await provider.getBalance(wallet.address);
    console.log("Wallet Balance:", ethers.formatEther(balance), "QBP");

    console.log("\nInteracting with Quantum Oracle...");
    const oracleAddress = "0x..."; // Deployed QuantumOracle contract address
    const oracle = new QuantumOracle(oracleAddress, wallet);

    // Example: Submit a quantum job
    try {
        console.log("Submitting a quantum job...");
        const tx = await oracle.submitJob(8, ethers.randomBytes(32), 10000, "50.0");
        console.log("Job submission transaction sent:", tx.hash);
        const receipt = await tx.wait();
        console.log("Transaction confirmed in block:", receipt.blockNumber);
    } catch (error) {
        console.error("Failed to submit job:", error.message);
    }

    // Example: Verify a post-quantum signature
    try {
        console.log("\nVerifying an ML-DSA signature on-chain...");
        const message = "Hello, Quantum World!";
        const messageHash = ethers.id(message);

        const signature = pqcrypto.sign(wallet.mldsaPrivateKey, ethers.getBytes(messageHash));
        const serializedSig = pqcrypto.serializeSignature(signature);

        const isValid = await oracle.verifyPostQuantumSignature(
            messageHash,
            pqcrypto.serializePublicKey(wallet.mldsaPublicKey),
            serializedSig
        );
        console.log("Is the signature valid?", isValid);
    } catch (error) {
        console.error("Failed to verify signature:", error.message);
    }
}

// Dummy pqcrypto library for demonstration.
// In the production implementation, this is a WebAssembly module
// compiled from the Go pqcrypto package (qbp-chain/crypto/pqcrypto),
// exposing FALCON-1024, ML-DSA-65, ML-KEM and SHA-999 to the browser.
// The Go WASM build is fully compatible with the Ethereum/go-ethereum stack.
const pqcrypto_lib = {
    generateMLDSAKeyPair: () => ({ secret: "...", public: "..." }),
    getPublicKey: (privKey) => privKey.public,
    serializePublicKey: (pubKey) => new Uint8Array(1952),
    sign: (privKey, hash) => ({ r: "...", s: "..." }),
    serializeSignature: (sig) => new Uint8Array(3309),
};

// Run example if the script is executed directly
if (require.main === module) {
    main().catch(console.error);
}
